Fix issue with consecutive updates, removing the option to return a promise in action creators

elisherer opened the pull request:
I Fixed the issue with consecutive action calls

I **removed** the option to return a `Promise` from action creators.
This seems unnecessary because you could now always run the action and it will apply on the current state, so running it from a promise created in the code (not in the store actions) would work the same.

I also added a unit test for it (which failed on the version before my change, see first commit).

**Point of thought**: Shouldn't we use the middleware as a pipeline to interfere with the action creators, passing them the functional setState and let the middleware decide if and when to pass it to the next middleware (allowing thunks/promise middlewares like in redux).

didierfranc commented:
Seems great I'll try it this night ;)

nlucero commented on src/index.js:
```diff
@@ -43,31 +43,35 @@ const createStore: CreateStore = (
 
-  const setState: CustomSetState = (action, state, ...args) =>
+  const setState: CustomSetState = (action, ...args) =>
     new Promise(resolve => {
```
Do we still need to return a `Promise` here?

didierfranc commented:
What about that kind of action ?

```js
const fetchSomething = async (state) => {
  const response = await fetch().then(response => response.json())
  return response.data
}
```

didierfranc commented:
What about that kind of action ?

```js
const fetchSomething = async (state) => {
  const response = await fetch(`.../${state.user.id}`).then(response => response.json())
  return response.data
}
```

elisherer commented:
Do it like:
```js
const response = await fetch(`.../${user.id}`).then(response => response.json())
actions.setResponseData(response)
```
The current state is available to invoke the async call. and on response you could invoke the action. I don't see a disability in this way.

didierfranc commented:
It's less convenient what do you think of this test https://github.com/didierfranc/react-waterfall/commit/647bbd4d74f7c37137f1f64c5b3dc65a009ffa41

The support for async function action can't be removed it's one of the most used thing with react-waterfall. Why don't we keep both ? Supporting multiple actions call and also async things seems great to me.

didierfranc commented:
It's less convenient what do you think of this test https://github.com/didierfranc/react-waterfall/commit/647bbd4d74f7c37137f1f64c5b3dc65a009ffa41

The support for async function action can't be removed it's one of the most used thing with react-waterfall. Why don't we keep both ? Supporting consecutive actions call and also async things seems great to me.

ktunkiewicz commented:
There's nothing bad in this project being "more and more" like redux. I'm actually looking at this project as a lightweight replacement for Redux.

I'm currently working on creating some lightweight, React based widgets, to make the code at least similar to what we write in our main application where we use React + Redux + a lot lot of other libraries. I use `react-waterfall` in conjunction with `preact` and `preact-promise` which gives me a few kilobytes base which is multiple times smaller footprint than React+Redux+all the boilerplate required.

What i'm missing in 'react-waterfall' library is the ability to asynchronously change state from inside the action "the redux way". Or is it possible but 

I imagine something like this would work ideally for use cases I usually encounter:
```
actionsCreators: {
  getQuote: (amount, currency) => (state, store) => {
    store.setState({ isGettingQuote: true, quote: null  });
    api.getQuote({ amount, currency })
      .then((res) => {
        store.setState({ isGettingQuote: false, quote: res.quote });
      })
      .catch((e) => {
        store.setState({ isGettingQuote: false})
        tools.parseApiError(e);
      });
  }
}
```

IMHO **having the state updated synchronously, and only by returning from an action, is an antipattern**. What this means is that you'll have to write your components like this:
```
<button
  onClick={(e) => {
    this.setLoading(true);
    this.getQuote(this.state.amount, this.state.currency);
    // how do I even check success here??
    this.setLoading(false);
  }}
>getQuote</a>
```

Actions should be the only place where logic happens. The components should only gather some data and call an action.
Then that action must do it's job, and the result of that action should be a change in state - in some time, in some way, not necessarily synchronously (because it may be calling an API, loading some data or doing something else) and the result of an action may be different depends on API response, etc. 

But for this, action creators must have access to setState.

elisherer commented:
We could maybe change the api to be promise creators which will resolve to setState function.

i.e.
```js
actionName: () => state => ..newState..
```

That way we could run the promise creator and see what we get.
if an object or a function then run straight to setState. if a promise, then resolve its value into setState (that's the most flexible solution I can think of)

elisherer commented:
@didierfranc , OK, so I implemented the middleware as discussed, (resembles redux almost completely), I put two examples in the unit tests (one for thunk, allowing the async action creators, and one for API calls inside a middleware).
I changed the middleware API to almost the same as redux with some extras.
Hope you appreciate the change :smile:

didierfranc commented:
I'm afraid to bring redux complexity (middleware, thunk ...) to react-waterfall with that kind of changes what do you think ?

elisherer commented:
In order to have the same flexibility you have with redux and really have it as a worthy in-place replacement (the way I used it until now), you need to have a very easy api for the normal developer and somewhat advanced api for the advanced developer. I didn't see a way to get passed the limitation of inability to pass async functions to setState.

ktunkiewicz commented:
Hi guys, thanks for including me in the discussion!
I've been experimenting with chainging the way actions work, to make it work in my use case. My goal is to make this a good replacement for Redux.

What works best for me, so far, is this construction:
```
actionName: (actionParameters) => (state, actions) => { /* some code */ }, 
```

and the if the action creator returns:
- an *object* - it is applied straight to state. This way very simple actions can be constructed, and behave synchronously. Example:
```
    setLoading: (isLoading: boolean) => ({ loading }) => ({
      loading: isLoading,
    }),
```

This is the current behaviour, allow simple, synchronous state changes.

- a *void* (or basically anything that's non-object) - The action will handle updating state/calling other actions itself. Example:
```
  fetchQuoteStart: (amount, currency) => (state, actions)  => {
    // getting some common info from global state, that is required in API call params
    const { mode, customerId } = state;
    // make some API call
    Api.fetchQuote({
      mode,
      customerId,
      amount,
      currency
    })
      .then((response) => {
        // on success we want to trigger some action
        actions.fetchQuoteSuccess(response);
      })
      .catch((err) => {
        // on failure we want to trigger some action
        actions.parseValidationErrors(err);
      })
  }
```

The reason why I put `(state, actions)` in the thunk, instead of just `(state)`, is that then all actions taken will be properly passed through middlewares. The behaviour will also be more predictable.
It should be then discouraged in documentation to change state directly from action and rather to make it a separate action that returns partial state. IMHO this kind of side-effect-free actions are much easier to maintain.
Additionally, with this approach, the Redux inspector plugin will show each separate action taking place, just like with Redux's dispatch. So it is easy to see what's going on when debugging.

I'm also thinking about what to do when a promise is returned. I don't see a good use case for "promise returned from action changing a state". Althought I do see a use case, that we used a lot in our code, to "return Promises from action creator for chaining purposes".

Example:
```
actionsCreators:
  initCalculator: () => (state, actions) => {
    actions.setLoading(true);
    Promise.all([
      actions.fetchCurrencyList(),
      actions.fetchAvailableLanguages(),
    )
    .then(() => {
      // loaded all stuff required to show UI
      actions.setLoading(false);
    })
    .catch((err) => {
      // could not initialise application
      actions.showGlobalError();
    })
```

In the above example, the successful loading will be seen like this in Redux devtools:
```
setLoading   // true
fetchCurrencyList
fetchAvailableLanguages
fetchCurrencyListSuccess   // not included in the above example, it was executed by fetchCurrencyList
fetchAvailableLanguagesSuccess // not included in the above example, it was executed by fetchAvailableLanguages
setLoading  // false
```

These are the changes I made to make it work like this:
https://github.com/didierfranc/react-waterfall/compare/master...ktunkiewicz:actions_enhanced?expand=1#diff-1fdf421c05c1140f6d71444ea2b27638R66

Also this change allows executing actions from Redux devtools:
https://github.com/didierfranc/react-waterfall/compare/master...ktunkiewicz:actions_enhanced?expand=1#diff-dd92fe9d8fb8e8088bb0eaae3244cce2R36

My branch has lots changes in it because I also fixed all linting errors, and going to also fix all flow typing. My goal is to have it typehinted in a way that will allow type checking at React components level.

And a bonus fact: `react-waterfall` works with `preact` :)
A little trick is required to make it happen, but it's nothing hacky. Because `preact` doesn't come with React Contex API, and it is a separate `preact-context` library, you have to create a small proxy library that import both `preact` and `preact-context` and re-exports both but joined.

So my plan of susbituting React with Preact and Redux with React-Waterfall works good so far.

ktunkiewicz commented:
My approach described above isn't actually redux-like, it is a bit of a shortcut, but it does the job. Reading coded from Elisherer now to see if it could do the same work.

didierfranc commented:
I published a new beta for the v4 https://github.com/didierfranc/react-waterfall/releases/tag/v4.0.0-beta.3

didierfranc commented:
I published a [new beta](https://github.com/didierfranc/react-waterfall/releases/tag/v4.0.0-beta.3) for the v4.

This pull request and the above discussion addresses some great problematics but implies a lot of changes I don't know how to do with it, what do you think @elisherer @ktunkiewicz @nlucero ?

didierfranc commented:
I published a [new beta](https://github.com/didierfranc/react-waterfall/releases/tag/v4.0.0-beta.3) for the v4.

This pull request and the above discussion addresses some great problematics but implies a lot of changes I don't know what to do with it, what do you think @elisherer @ktunkiewicz @nlucero ?

elisherer commented:
As I said before, I have a problem with imposing actions to allow async state changes while the infrastructure being used (React's comoponent state) doesn't support it. Unless this will be solved, you will always have to explain (for example) why you can't just run normally two actions sequentially (which seems odd).
As to my solution, if you will embed the middleware I put in the unit test it will allow async actions (but in that case async actions would still get the unchanged state as parameter).
To conclude my opinion, because I'm a redux purist 😁, I think state management and async procedures should be separate.
A developer can easily create an async wrapper around sync actions (or, not bad idea at all, write that code inside their component, a component can always replace a middleware).
Hope I helped 😇

didierfranc commented:
Yes it is a very inspiring discussions, I'm really happy to have published that few lines of code (react-waterfall itself), we discover good and bad things around it, and personally learned some thing.

FYI, I'm thinking about moving from component state to something more flexible.

Am i right if I said that moving to functional setState will remove all batching/optimisation from setState ?

elisherer commented:
Nope,
https://reactjs.org/docs/react-component.html#setstate
"Subsequent calls will override values from previous calls in the same cycle, so the quantity will only be incremented once. If the next state depends on the previous state, we recommend using the updater function form"

klassicd commented:
It hasn't been super useful to be able to await actions from a component because:
- componentDidMount and componentDidUpdate both expect `void` return and throw type errors when trying to make them async and await actions in them.
- The action is usually responsible for processing API errors and updating the store.   If you `await actions.getFoo()` in your component and it fails, you will still get an resolved(undefined) promise indicating a success, unless you explicitly return a rejected promise.   I then have to use try/catch blocks in my component to prevent other actions from running, and insure any "Uncaught Errors" don't block logic.  It's easier at this point to just connect the store. 

I really like the idea of being able to call actions from other actions.  I don't think the actions property should come before the arguments (as in v4.0.0-beta.3).  The more common use case is working from `(state, arguments)` and not `(state, actions, arguments)`.

didierfranc commented:
This is not possible since arguments is an array

klassicd commented:
What about `const finalArgs = [...args, actions];` and then `(state, ...finalArgs)`?
